073-improve-handling-of-CA-renewals-and-replacements-in-client-based-operands.md

```diff
@@ -0,0 +1,210 @@
+   At this point in time, the Kafka brokers still use the old server certificates signed by the old CA.
+3. The Kafka client based operand (e.g. Kafka Connect cluster) using the new `.spec.tls` configuration:
+   ```yaml
+   tls:
+      trustedCertificates:
+         - secretName: my-cluster-cluster-ca-cert
+           extension: crt
+   ```
+   detects that the `ca.crt` in the Secret was updated and a new `ca-<timestamp>.crt` added, triggering a rolling update of the operand to use the new CA.
+   After the rolling update, it trusts both the old and new CAs and continues working without any problems while the CA is being replaced.
+4. After all brokers are rolled to trust the new CA, another rolling update happens and new server certificates signed by the new CA are rolled out to the Kafka brokers in another rolling update.
+5. Once all brokers are rolled once again and use the new CA, the operand such as Kafka Connect will be able to connect again and start working.
+   The impact and length of the outage will depend on many different aspects:
+   * How many nodes does the operand such as Kafka Connect have.
+     If it is only one replica, the whole operand will be down.
+     With multiple replicas, only one replica should be down.
+   * How big and busy the Kafka cluster is.
+     A small cluster might take few minutes to move to the new CA completely.
+     A big cluster with many nodes that are busy and need more time to re-sync between restarts might take much longer.
+6. After all brokers (and other components such as ZooKeeper, Topic and User Operator, Cruise Control etc.) use the new CA, another rolling update will be done to completely remove the trust to the old CA by removing the earlier `ca-<timestamp>.crt` from the Secret and rolling all components once again.
+   This last rolling update does not impact operands such as Kafka Connect.
+7. The removal of `ca-<timestamp>.crt` from the Secret will cause rolling update of the client-based operand (e.g. Connect).
+   This rolling update will remove the trust to the old CA from the operand as well.
+   This is an additional rolling update that happens when the new API is used.
+
+### Examples
+
+The examples for the operands based on Kafka clients will be updated to use the new field:
+```yaml
+  tls:
+    trustedCertificates:
+      - secretName: my-cluster-cluster-ca-cert
+        extension: crt
+```
+
+This will make sure that when used with a Strimzi based cluster, everything - including CA replacement - will work without any outages and issues.
+
+## Affected projects
+
+This proposal affects only the Strimzi Cluster Operator project.
+
+## Backwards compatibility
+
+There is no impact on backwards compatibility.
+This proposal does not remove any Strimzi APIs.
+It only extends them.
+All existing custom resources will work without any change.
+
+## Rejected alternatives
+
+### Using a regular expression instead of extension
```

Comment:
I think this looks nice in theory and I agree it would be more understandable for users. But the Java API you link to seems to be designed to match actual filesystems and paths. Not keys in a HashMap. So I'm not sure this is a good idea to use that. There are also other implementations such as in [Apache Commons](https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/FilenameUtils.html#wildcardMatch(java.lang.String,java.lang.String)). But not sure those are worth the additional dependency. Or did I miss some part of the Java API not dependent on filesystem?

Reply:
It's true that the Java API is in terms of the `Path` class, but there's nothing coupling it to any actual path on any actual filesystem. I.e. you could write the function:

```java
boolean matches(String input, String glob) {
    return FileSystems.getDefault().getPathMatcher("glob:" + glob).matches(Paths.get(input));
}
```

and that function would do what you need without having access to the the real filesystem. 

I agree that adding a dependency on Apache Commons _just for this_ doesn't seem like a great tradeoff.

Reply:
Ok, I incorporated it. Please have another look. I wonder if calling it `pattern` would be better than `certificatesMatching` => less space for typos, shorter, easier to understand maybe? What do you think?

Reply:
I don't have a very strong opinion about the name. `pattern` is _ok_ (it is shorter), but it doesn't convey:

* the pattern syntax (people might think it's a regexp)
* what the pattern will be used to match against (the keys of the entries in the Secret)
* what expectations there are about what the matches need to _be_, namely certificates (but maybe this isn't really needed since it's nested in a thing called `trustedCertificates`).

The first two of those criticisms also apply to `certificatesMatching`. But I don't think something like `keyGlob` is better, so unless someone can think of a better name I think `pattern` would suffice.

Reply:
Ok, let's see if others will have some thoughts about it.

Reply:
Tbh I think `pattern` is better. We already have `certificates` in the root field (referring to `trustedCertificates`) and other than that, I think the pattern term is well recognized and understandable.

Reply:
I also think `pattern` is better than `certificatesMatching`,  if we wanted to be very explicit another option would be `filenamesGlob`, or `filenamePattern` to make it clear what the pattern is for? I know technically it's a key in a secret not a filename, but given it has a filename extension I think it would be clearer to the user what it's referring to than using `keyGlob/Pattern`

Reply:
Ok, let's go with `pattern`. I will update the proposal.